In [14]:
import numpy as np
import matplotlib.pyplot as pyplot
import scipy.stats as sps
from sklearn import datasets,cross_validation,neighbors,metrics
from matplotlib.colors import ListedColormap
from sklearn.naive_bayes import BernoulliNB, MultinomialNB, GaussianNB
%matplotlib inline

In [4]:
digits_data = datasets.load_digits()
digits_data = [digits_data['data'],digits_data['target']]

breast_cancer_data = datasets.load_breast_cancer()
breast_cancer_data = [breast_cancer_data['data'],breast_cancer_data['target']]

In [12]:
print(digits_data[0][:2])

[[  0.   0.   5.  13.   9.   1.   0.   0.   0.   0.  13.  15.  10.  15.
    5.   0.   0.   3.  15.   2.   0.  11.   8.   0.   0.   4.  12.   0.
    0.   8.   8.   0.   0.   5.   8.   0.   0.   9.   8.   0.   0.   4.
   11.   0.   1.  12.   7.   0.   0.   2.  14.   5.  10.  12.   0.   0.
    0.   0.   6.  13.  10.   0.   0.   0.]
 [  0.   0.   0.  12.  13.   5.   0.   0.   0.   0.   0.  11.  16.   9.
    0.   0.   0.   0.   3.  15.  16.   6.   0.   0.   0.   7.  15.  16.
   16.   2.   0.   0.   0.   0.   1.  16.  16.   3.   0.   0.   0.   0.
    1.  16.  16.   6.   0.   0.   0.   0.   1.  16.  16.   6.   0.   0.
    0.   0.   0.  11.  16.  10.   0.   0.]]


In [13]:
print(breast_cancer_data[0][:2])

[[  1.79900000e+01   1.03800000e+01   1.22800000e+02   1.00100000e+03
    1.18400000e-01   2.77600000e-01   3.00100000e-01   1.47100000e-01
    2.41900000e-01   7.87100000e-02   1.09500000e+00   9.05300000e-01
    8.58900000e+00   1.53400000e+02   6.39900000e-03   4.90400000e-02
    5.37300000e-02   1.58700000e-02   3.00300000e-02   6.19300000e-03
    2.53800000e+01   1.73300000e+01   1.84600000e+02   2.01900000e+03
    1.62200000e-01   6.65600000e-01   7.11900000e-01   2.65400000e-01
    4.60100000e-01   1.18900000e-01]
 [  2.05700000e+01   1.77700000e+01   1.32900000e+02   1.32600000e+03
    8.47400000e-02   7.86400000e-02   8.69000000e-02   7.01700000e-02
    1.81200000e-01   5.66700000e-02   5.43500000e-01   7.33900000e-01
    3.39800000e+00   7.40800000e+01   5.22500000e-03   1.30800000e-02
    1.86000000e-02   1.34000000e-02   1.38900000e-02   3.53200000e-03
    2.49900000e+01   2.34100000e+01   1.58800000e+02   1.95600000e+03
    1.23800000e-01   1.86600000e-01   2.41600000e-01 

In [16]:
def get_score(data,target):
    models = [BernoulliNB(),MultinomialNB(),GaussianNB()]
    scores = []
    for model in models:
        scores.append(cross_validation.cross_val_score(model,data,target).mean())
    return scores

In [17]:
get_score(digits_data[0],digits_data[1])

[0.82582365077805819, 0.87087714897350532, 0.81860038035501381]

Лучшую точность для digits дает MultinomialNB (0.87), что логично, т.к. признаки - целые числа в некотором диапазоне.

In [18]:
get_score(breast_cancer_data[0],breast_cancer_data[1])

[0.62742040285899936, 0.89457904019307521, 0.9367492806089297]

Для breast_cancer_data лучше всех подошел GaussianNB (0.94), что ожидаемо, т.к. признаки - вещественные числа.

### 1,2 :
- На digits : MultinomialNB (0.87)
- На breast_cancer_data : GaussianNB (0.94)

### 3 :
- False
- False
- False (в данных тестах Multinomial показало себя лучше на вещественных признаках) 
- True